src/main/java/net/fabricmc/loom/configuration/providers/minecraft/MergedMinecraftProvider.java

```diff
@@ -48,50 +48,73 @@ protected void initFiles() {
 
 	@Override
 	public List<Path> getMinecraftJars() {
-		return List.of(minecraftMergedJar);
+		return canMergeJars() ? List.of(minecraftMergedJar) : List.of(getClientJar(), getServerJar());
```

Comment:
This doesnt feel right, you would always expect `MergedMinecraftProvider` to have a single jar.

Reply:
That would mean adding a new MinecraftJarConfiguration, right? I'll have to make another attempt at that, the first time I tried it it seemed like I'd have to rewrite a bunch of stuff for it. But it's been a while so I can't remember the specific problems I ran into.

Reply:
Hold off for a bit, im cleaning some stuff up to help with this. 👍

Reply:
Ah I remember the issue I ran into when creating a new jar configuration for this. The jar configuration is used to create the minecraft provider, which in turn retrieves and parses the minecraft info, which is used to check if a version can be merged or not. Basically, it would force the user to manually choose which 'kind' of merged they need while there is no reason loom couldn't figure that out for itself. But for that the minecraft info needs to be parsed before the jar configuration is created.

Reply:
First thing I have done is remove that mess of generics: https://github.com/FabricMC/fabric-loom/pull/1070 makes things a lot easier :D (I hope)

I have then created a rough outline of how I think this could work here: https://github.com/modmuss50/fabric-loom-1/commit/32df745660b70803c81a4003aa786eecd2ad2181

I have created some new minecraft and intermediary providers that delgate to the existing single jar providers for the client and server. This basically means you end up with both the client and server jar as if you were using `clientOnlyMinecraftJar` and `serverOnlyMinecraftJar`. You end with with 2 seperate named client/server jars, the special handling to merge them can then happen in `NamedMinecraftProvider.LegacyMergedImpl`

Let me know what you think, I havent tried to run it partly becuase I dont have mappings to test with, and partly becuase you might notice something tottaly wrong with it 😆 

>  it would force the user to manually choose which 'kind' of merged they need while there is no reason loom couldn't figure that out for itself.

This is a valid issue, and still a problem with what I have done above. Ill look into seeing what could be done about that. It likely requires moving the `MinecraftMetadataProvider` out of the `MinecraftProvider`, I think this can be handled separately. Its not the end of the world for it to be specified for now.

Reply:
> First thing I have done is remove that mess of generics: #1070 makes things a lot easier :D (I hope)

Neat, that should make it less of a head ache for sure!

> Let me know what you think, I havent tried to run it partly becuase I dont have mappings to test with, and partly becuase you might notice something tottaly wrong with it 😆

I do have some questions about it, mostly 'cause I'm not familiar enough with Loom's inner workings.

It is not an issue that the intermediary and named providers provide separate client and server jars instead of one merged jar? Or are these jars otherwise never exposed and used anywhere else? If so, could the merging then be done at the start of the processed named provider step?

And something that feels more like a nitpick - is there a reason behind the use of delegate providers rather than passing in the client and server jars as defined in `MinecraftProvider`?

> This is a valid issue, and still a problem with what I have done above. Ill look into seeing what could be done about that. It likely requires moving the `MinecraftMetadataProvider` out of the `MinecraftProvider`, I think this can be handled separately. Its not the end of the world for it to be specified for now.

Fair enough, I'll look out for that then!

Reply:
> It is not an issue that the intermediary and named providers provide separate client and server jars instead of one merged jar?

Yes, I think that might be an issue. I forgot that you can still create a merged intermediary jar, thats easily solved though and would follow the same pattern as the named provider.

> And something that feels more like a nitpick - is there a reason behind the use of delegate providers rather than passing in the client and server jars as defined in MinecraftProvider?

Good point, that is a waste of time, you could use the same provider.